Fix sphere positions and light direction in sphere visual.

ron-adam opened the pull request:
Fixes to lighting and positioning for datoviz sphere visual.

[fixes for sphere visual - Issue 79](https://github.com/datoviz/datoviz/issues/79)

The cause of the positioning problem is the transform function converts from opengl convention to vulkan.  But the z axis is reversed.

This PR flips the z axis before it's called.  Line 31 in the `src/scene/glsl/graphics_sphere.vert` file.

In the transform function is located in common.glsl, and calls the to_vulkan() function to transpose from opengl convention to vulkan. 

Recommend to rename (or copy) the to_vulkan function to gl_to_vulkan_clip, and add a dvz_to_vulkan_clip that inverts the z axis. This makes it clear what the result will be and help avoid it being used in the wrong context.

The light direction problem is due to the light position not being corrected from dvz to vulkan.  The z axis is correct, but the y axis is reversed.   The light positioning problem was also present due to comparing world scale coordinates for the light, and view coordinates for the sphere.  

The remaining changes are present to ensure the correct values get to the correct places.  The conversions are done in the vertex shader and then passed to the fragment shader.  That may be slightly more efficient.

This is a draft pull request because I think it can be improved further.

rossant commented:
Many thanks for the careful test case, the investigation, and the proposed fixes! Could you post a screenshot of the same example, with the fixes in?

ron-adam commented:
I added the screen shot to the issue.

When I have a pull request, should I just put everything here, or is having the problem explained in the issue tracker good too?   I can do either or both.

> Recommend to rename (or copy) the to_vulkan function to gl_to_vulkan_clip, and add a dvz_to_vulkan_clip
> that inverts the z axis. This makes it clear what the result will be and help avoid it being used in the wrong context.

This suggestion isn't quite correct.   I mixed up the usage with the transform function.  So the suggestion should be to add functions gl_to_vulkan and dvz_to_vulkan, without the _clip at the end.

I don't know if the tranform should be changed or a new function created with a different name.   Changing it may effect other parts, a new function with a different name can avoid that.

rossant commented:
> When I have a pull request, should I just put everything here, or is having the problem explained in the issue tracker good too? I can do either or both.

I think we could have the problem description/test case as an issue, and a proposed fix to the issue as a pull request (with a reference to the issue), exactly as you did here!

ron-adam commented:
I believe I got this sorted out now.

1.  Removed the to_vulkan() call from the tranform function.

I believe this was in the wrong place and was doing the wrong conversion for what we need.

2.  Created a `dvz_to_vulkan_coord(pos)`  function in common.glsl.

Any coordinates received are first converted to vulkan.  This could also be placed in the C files outside the shaders.
(Thoughts?)

3.  Reversed the y mouse direction in the arcball.c.  (It was negative, so this puts it back to positive.)


With these changes, you first use the dvz_to_vulkan_coord(pos) for any incoming dvz coordinates,  then everything works correctly in vulkan coordinates.

This means other visuals will need to be updated as well.  But after words, they should all work in the correct dvz coordinates and in the correct directions. 

I'm going to do a bit more testing to see how this effects the other visuals to be sure it's a good fix.  They should be fairly easy to correct.  But It may be better to do each one as a separate issue.

The main() function in the vertex shader looks like this after these changes.   The fragment shader is nicer too.  Moving the coordinate conversion into the calling C files would could clean it up more.  But it may be slower for data being updated often.
```
void main()
{
    // Convert coordinates
    out_pos = dvz_to_vulkan_coord(pos);
    out_light_pos = dvz_to_vulkan_coord(params.light_pos.xyz);

    out_world_pos = mvp.model * out_pos;
    out_view_pos = mvp.view * mvp.model * out_pos;
    out_radius = radius;
    out_color = color;

    // Project the position to clip space using the transform function
    gl_Position = transform(out_pos.xyz);

    // Set the point size to the diameter of the sphere in pixels
    float distance_to_camera = length(out_view_pos.xyz);
    gl_PointSize = (2.0 * radius) / distance_to_camera;
}
```

rossant commented:
Thanks. One constraint is that the cglm module assumes an OpenGL coordinate system by default, which was complicating things. However, I just saw that we can [change the settings](https://cglm.readthedocs.io/en/latest/opt.html#clipspace-option-s).

I think it's best if we can keep the CPU data in Datoviz coordinates, and we handle the Vulkan coordinate system transformation in the shaders.

ron-adam commented:
I updated the pull request to include the (more obvious) needed changes elsewhere.  It was fairly straight forward.  

It still needs a bit of clean up.  (Removing some commented out lines.)

With this,  it also corrects the camera rotation directions.   Considering a point between you and the camera,  +1 around the x axis will move the point up.  +1 around the y axis will move the point to the right,  And +1 around the z axis will rotate the point clockwise.  (If it isn't in the center.)   I believe this corresponds with mouse movements also.

There are probably other places that will need to be adjusted, but I don't think many.   

I don't mind if you point out minor adjustments, or style related changes.  Or any other aspects.  I try to stay consistent with the existing code, but I don't know enough about the projects code base yet and the specific roadmap goals.

As an experiment, I moved the lighting code to a function in the sphere fragment shader to see what was involved.  It seems to me that the lighting code will be used in various places.  Possibly functions related to it can be placed in a render.glsl file.  Other thoughts are to seperate material (the ambient, diffuse, and specular properties belonging to the surface), from the lights, with directions, locations, and colors.   (Sometime later.)

rossant commented:
Great, thanks! I'll have a look soon.

> With this, it also corrects the camera rotation directions. Considering a point between you and the camera, +1 around the x axis will move the point up. +1 around the y axis will move the point to the right, And +1 around the z axis will rotate the point clockwise. (If it isn't in the center.) I believe this corresponds with mouse movements also.

I'm not sure I understand this, could you clarify?

rossant commented:
To be clear, currently, the arcball does *not* rotate the camera, it rotates the object (it only affects the model matrix, not the view nor projection matrix). So the lights etc should remain at their exact position while you're rotating the object around its center.

In an earlier version of Datoviz, I had this arcball but also another interaction pattern which was moving the camera instead. The new version should definitely also provide a fly camera and/or a "first-person-like" camera. I can find the code I had from this earlier version if needed.

rossant commented:
> I think we need to get this correct if datoviz is to be used for math and science visualizations.

Agreed!

> Looking along the z axis. (into the screen), increasing the z angle will rotate the arcball clockwise.

You mean that z+ is towards the person in front of the monitor, whereas z- is away from them, right? If so, yes, I agree this is the wanted behavior!

> I also have a spin camera that I wrote from the Vispy arcball camera. With it you can spin the objects in any direction and it keeps spinning until you stop it. The speed is controlled by how fast you drag it with the mouse.

Nice! I'd love to have that in the C API.

By the way, I'd like to point out for future reference that math and computer graphics do not use the same axis convention, which is a shame, but there's nothing we can do about it.

These are the math conventions, from [Wikipedia](https://en.wikipedia.org/wiki/File:Gimbaleuler2.svg): (look at the x, y, z blue axes)

![](https://upload.wikimedia.org/wikipedia/commons/thumb/3/34/Gimbaleuler2.svg/798px-Gimbaleuler2.svg.png)

Whereas  this is the [OpenGL convention](https://learnopengl.com/Getting-started/Coordinate-Systems) that Datoviz is also using (right?):

![](https://learnopengl.com/img/getting-started/coordinate_systems_right_handed.png)

See also this figure [from this page](https://caiorss.github.io/C-Cpp-Notes/computer-graphics.html):

![](https://github.com/user-attachments/assets/aeb2b533-6c57-4bc2-a2d4-00dd0c750691)

Fortunately, both are right-handed, meaning you can go from one to another with a rotation in 3D. It also means that if you imagine being "spiked" on the z axis while looking at the x and y axis, to go from x+ to y+, you have to rotate around the z axis in the *anticlockwise* direction.

I think it's clearer if we "forget" about the Vulkan convention in the documentation, which is yet another convention (!), and hide it as much as possible in the shaders...

ron-adam commented:
> You mean that z+ is towards the person in front of the monitor, whereas z- is away from them, right? If so, yes, I agree this is the wanted behavior!

> Whereas this is the [OpenGL convention](https://learnopengl.com/Getting-started/Coordinate-Systems) that Datoviz is also using (right?):

Ok, I got the impression datoviz's z was increasing into the distance as seen through the monitor.    That makes sense for some things, but the z axis is then reversed compared to OpenGL.

From the datoviz userguide:
> The coordinate system is defined as follows:
> 
> x: left to right [-1, +1]
> y: bottom to top [-1, +1]
> z: front to back [-1, +1]
> Positions must be provided in a normalized coordinate system, known as normalized device coordinates (NDC) in computer graphics terminology. Since your data is typically not in this range, you'll need to manually normalize it to the [-1, +1] interval before passing it to Datoviz.
> 
> Datoviz v0.2 does not yet include built-in axes or data normalization features, but these will be introduced in v0.3.

Possibly front to back is unclear,  looks like I had the meaning of front and back reversed?   Looking at an object (or triangles faces) in the monitor, the front is towards us, and the back is away from us.  `z: front to back [-1, +1]` would place negative positions behind the camera in that context.   (But this is not the correct context.)

So you want to use the Opengl convention for datoviz coordinates?

Example:   To light up the front of an object, the coordinates needed to place the light behind the camera would be (0, 0, +1)  if the camera is at (0, 0, 0), and the object at (0, 0, -1).

And the rotation looking towards the "negative" values will be counterclockwise for positive z rotation values.  It would be clockwise if z was increasing into the distance, but opengl values decrease in the distance.  (As seen through the monitor.)  I believe this is the current dev branch behavior.

If we are now on the same page, I can update the pull request accordingly.  It possibly will be much reduced.

ron-adam commented:
Yes, counterclockwise.   I believe we are in agreement now.

As to what the convention should be for datoviz coordinates,  as a python programmer, I think what is more consistent with other python math libraries. 

A quick google returns...

> In most Python math libraries and applications, the default coordinate system is right-handed; meaning that the positive x-axis points right, the positive y-axis points up, and the positive z-axis points outwards, following the "right-hand rule" convention. 

So I think we are on the correct path forward.

ron-adam commented:
Ok... The files are updated.  It's a much smaller patch now and only affects the sphere vertex and fragment shader.

I made some changes to the depth calculation and lighting in the fragment shader.
The depth calculation is moved in front of the lighting calculations so the lighting uses the updated position. 
```
    // Update depth
    vec4 view_pos = in_eye_pos;
    float d = length(view_pos * viewport.size.x);
    view_pos.xyz += (in_radius * normal)/d;
    gl_FragDepth = 1.0 - 1.0/(1.0 + length(view_pos));
```
This has better behavior when zooming into the arcball.

The lighting changes result in nicer gradients on the arcball and feathering out the specular lighting on the very edge of the sphere.  It's noticeable if the light is behind the spheres.  It also acts a bit like a anti-alias effect.

Before:
![image](https://github.com/user-attachments/assets/43c7baaa-01ac-475e-a092-18bcdd357f74)

After:
![image](https://github.com/user-attachments/assets/76d4f8df-55f6-4afa-bb32-4b1997be7ebb)
![image](https://github.com/user-attachments/assets/af3c26bc-ceab-4079-a7d8-807b8b4e412a)
![image](https://github.com/user-attachments/assets/86d8b78d-7330-4483-8f76-678f310ad5f6)
![image](https://github.com/user-attachments/assets/87fe07c6-adb9-458c-b42a-97eff38455ba)
![image](https://github.com/user-attachments/assets/56e9ee68-72d0-4caa-8b7b-80fa9a1e2d14)

rossant commented:
Wow, great job, thanks! It looks so much nicer now! It's ready to merge?

ron-adam commented:
Thanks.. I want to verify the depth calculations a bit more.  But it's close.

Also while compiling this I kept getting a Change sign warnings, which slowed down compiling quite a bit.

Adding type casts in `include/datoviz/_time_utils.h` line 158 resolved it.
``` 
uint64_t nanoseconds = (long long unsigned int)(counter.QuadPart * 1000000000) / (long long unsigned int)frequency.QuadPart;
```

I can include that in the pull request, or another, or you can do it.  (It's just one line.)

ron-adam commented:
Ok, it's ready.
